### Script purpose

- Replicate and compare with: Anastasopoulos, L. Jason, and Andrew B. Whitford. 2019. “Machine Learning for Public Administration Research, With Application to Organizational Reputation.” Journal of Public Administration Research and Theory 29 (3): 491–510. https://doi.org/10.1093/jopart/muy060.
- Use glove-wiki-gigaword-100 as embedding layer.

In [1]:
# Ignore warnings.
import warnings
warnings.filterwarnings("ignore")

# obtain reproducible results
import numpy as np
import tensorflow as tf
import random as rn
from keras import backend as K

# The below is necessary for starting Numpy generated random numbers
# in a well-defined initial state.
np.random.seed(42)
# The below is necessary for starting core Python generated random numbers
# in a well-defined state.
rn.seed(12345)
# Force TensorFlow to use single thread.
# Multiple threads are a potential source of non-reproducible results.
# For further details, see: https://stackoverflow.com/questions/42022950/
session_conf = tf.ConfigProto(intra_op_parallelism_threads=1,
                              inter_op_parallelism_threads=1)
# The below tf.set_random_seed() will make random number generation
# in the TensorFlow backend have a well-defined initial state.
# For further details, see:
# https://www.tensorflow.org/api_docs/python/tf/set_random_seed
tf.set_random_seed(1234)
sess = tf.Session(graph=tf.get_default_graph(), config=session_conf)
K.set_session(sess)
# Rest of code follows ...

# Check GPU device.
print(K.tensorflow_backend._get_available_gpus())

Using TensorFlow backend.




['/job:localhost/replica:0/task:0/device:GPU:0']


In [2]:
#https://richliao.github.io/supervised/classification/2016/11/26/textclassifier-convolutional/
#https://blog.keras.io/using-pre-trained-word-embeddings-in-a-keras-model.html
#RNN

import os
import pandas as pd
import re
from tqdm import tqdm
from multiprocessing import Pool
from spellchecker import SpellChecker
import string
import math

import nltk
from nltk.stem import WordNetLemmatizer
from nltk.corpus import wordnet
from nltk.corpus import stopwords
import keras.preprocessing.text as kpt
from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences
from keras.layers import Embedding
from sklearn.model_selection import train_test_split
from nltk import word_tokenize
import matplotlib.pyplot as plt

# For encoding labels.
from sklearn.preprocessing import LabelEncoder
from keras.utils import np_utils

In [3]:
df_coded_tweet=pd.read_csv('../../dataset/muy060_suppl_supplementary_appendix/muy060_Appendix_2.csv')
df_coded_tweet.loc[df_coded_tweet.JasonCode==2, 'moral_bin']=1
df_coded_tweet.moral_bin.fillna('0', inplace=True)
df_coded_tweet['moral_bin']=[int(s) for s in df_coded_tweet.moral_bin]
df_coded_tweet.sample(5)

HITID  \
65   3AXFSPQOYQYNNLFNVXHLO2FHNRHFJ3   
114  3N5YJ55YXG3OAKP0ZFNL38L7ZV9ANP   
16   35A1YQPVFEGZQD2S73JCQP94LIEI5B   
141  3TX9T2ZCB91FYM6M38U7GKP71SJWZ2   
156  3GVPRXWRPHUEDHGBVWY9O9N0CO97IF   

                                                  Text             Answer1  \
65   RT NWSNewOrleans Our 6ft 6in MIC next to 11ft ...  None of the above.   
114  RT fema Storm surge can be one of the most dan...  None of the above.   
16   RT statedeptspox Thank you  Dr Alison Mann fro...  None of the above.   
141  CommerceGov announces preliminary countervaili...  None of the above.   
156  Whos a good boy?A USAirForce airman rewards Ni...    Moral Reputation   

                     Answer2 Agreement              Answer  \
65   Performative Reputation        No        no_agreement   
114     Technical Reputation        No        no_agreement   
16        None of the above.       Yes  None of the above.   
141  Performative Reputation        No        no_agreement   
156     Technical Reputation        No        no_agreement   

                        Date  JasonCode  moral_bin  
65   2018-07-06 18:03:42 UTC          0          0  
114  2018-07-06 17:59:34 UTC          2          1  
16   2018-07-06 17:59:48 UTC          0          0  
141  2018-07-06 18:04:45 UTC          1          0  
156  2018-07-06 18:08:55 UTC          2          1

In [4]:
df_agency_tweets_database=pd.read_csv('../../dataset/muy060_suppl_supplementary_appendix/muy060_Appendix_1.csv', encoding='ISO-8859-1')
df_agency_tweets_database.sample(3)

agency_id                                         tweet_text  \
10430  Interior  RT @SecretaryZinke: HUGE!  Latest @USGS and @B...   
11393    HUDgov  RT @HUDSoutheast: Celebrating #CollegeSigningD...   
17471    ENERGY  Former Energy Secretary @ErnestMoniz and @MIT ...   

       tweet_favorites  tweet_retweets  tweet_created  
10430                0             103   1.513968e+09  
11393                0               1   1.461698e+09  
17471               18              10   1.510787e+09

In [5]:
text_token_list_train=df_coded_tweet['Text']
text_token_list_pred=df_agency_tweets_database['tweet_text']

In [6]:
# Prepare labels
from sklearn import preprocessing
lb = preprocessing.LabelBinarizer()

lb.fit(df_coded_tweet.moral_bin.unique())
y_train=lb.transform(df_coded_tweet['moral_bin'])
from keras.utils import to_categorical
y_train = to_categorical(y_train)

In [7]:
# Build word index for train and validation texts.
tokenizer=Tokenizer()
tokenizer.fit_on_texts(text_token_list_train.to_list()+text_token_list_pred.to_list()) # Including uncoded tweets, vocabulary must be comprehensive.
print(list(tokenizer.word_index.items())[0:5])

[('https', 1), ('t', 2), ('co', 3), ('the', 4), ('to', 5)]


In [8]:
## Train dataset.
# Text to sequences.
seq_encoding_text_train=tokenizer.texts_to_sequences(text_token_list_train)
seq_encoding_text_pred=tokenizer.texts_to_sequences(text_token_list_pred)

# Pads sequences to the same length (i.e., prepare matrix).
x_train=pad_sequences(sequences=seq_encoding_text_train,
                      maxlen=max([len(s) for s in seq_encoding_text_train+seq_encoding_text_pred]), # Max length of the sequence. Must consider prediction dataset.
                      dtype = "int32", padding = "post", truncating = "post", 
                      value = 0 # Zero is used for representing None or Unknown.
                     )
x_pred=pad_sequences(sequences=seq_encoding_text_pred,
                      maxlen=max([len(s) for s in seq_encoding_text_train+seq_encoding_text_pred]), # Max length of the sequence. Must consider prediction dataset.
                      dtype = "int32", padding = "post", truncating = "post", 
                      value = 0 # Zero is used for representing None or Unknown.
                     )

In [9]:
import gensim.downloader as api
EMBEDDING_DIM=100
glove_word_vector=api.load('glove-wiki-gigaword-'+str(EMBEDDING_DIM))
# glove_word_vector=api.load('glove-twitter-200')

In [10]:
embedding_matrix = np.zeros((len(tokenizer.word_index)+1, EMBEDDING_DIM)) # Plus one: embedding matrix starts from 0, word index starts from 1.

In [11]:
for word, index in tokenizer.word_index.items():
    try:
        embedding_matrix[index] = glove_word_vector.get_vector(word)
    except:
        pass
        # words not found in embedding index will be all-zeros.

In [12]:
input_length=len(x_train[0])
# https://machinelearningmastery.com/use-word-embedding-layers-deep-learning-keras/
embedding_layer = Embedding(input_dim=len(tokenizer.word_index)+1, # Size of vocabulary.
                            input_length=input_length, # Length of input, i.e., length of padded sequence.
                            output_dim=EMBEDDING_DIM, # Size of the vector space in which words will be embedded.
                            weights=[embedding_matrix],
                            trainable=False)

In [13]:
from keras.models import Sequential
from keras.layers import Dense, Flatten, GlobalMaxPool1D, Conv1D
from datetime import datetime
import itertools

num_filters=[32, 64, 128]
kernel_size=[3,5,7]
conv_act=['sigmoid', 'softplus', 'tanh', 'softmax']
act_32=['sigmoid', 'softplus', 'tanh', 'softmax']
act_16=['sigmoid', 'softplus', 'tanh', 'softmax']
out_act=['sigmoid', 'softplus', 'tanh', 'softmax']
param_list=list(itertools.product(num_filters, kernel_size, conv_act, act_32, act_16, out_act))

In [ ]:
result_list=[]

for num_filters, kernel_size, conv_act, act_32, act_16, out_act in tqdm(param_list):
    with tf.device('/gpu:0'): # Small dataset, may even increase time because of data transfer between C/GPU.
        model = Sequential()
        model.add(embedding_layer)
        # model.add(Flatten())
        model.add(Conv1D(num_filters, kernel_size, activation=conv_act))
        model.add(GlobalMaxPool1D())
        model.add(Dense(units=32, activation=act_32))
        model.add(Dense(units=16, activation=act_16))
        model.add(Dense(units=len(y_train[0]), activation=out_act))
        # compile the model
        model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['acc'])
        # F1, precision, and recall removed. https://github.com/keras-team/keras/issues/5794
        # fit the model
        history=model.fit(x_train, y_train, validation_split=0.3, epochs=100, verbose=0)
    result_list+=[[max(history.history['val_acc']), [num_filters, kernel_size, conv_act, act_32, act_16, out_act]]]
    ############### Clean up and rebuid objects; otherwise, tensor graph becomes larger and larger and requires more time. ##########
    del model
    K.clear_session() # Will cause error: Tensor must from the same graph. Cause: Embedding layers changed. Solution: rebuild embedding layer.
    ##### Initialize session for reproducibility #####
    np.random.seed(42)
    rn.seed(12345)
    session_conf = tf.ConfigProto(intra_op_parallelism_threads=1,
                                  inter_op_parallelism_threads=1)
    tf.set_random_seed(1234)
    sess = tf.Session(graph=tf.get_default_graph(), config=session_conf)
    K.set_session(sess)
    ##################################################
    # Rebuid embedding layer.
    embedding_layer = Embedding(input_dim=len(tokenizer.word_index)+1, # Size of vocabulary.
                                input_length=input_length, # Length of input, i.e., length of padded sequence.
                                output_dim=EMBEDDING_DIM, # Size of the vector space in which words will be embedded.
                                weights=[embedding_matrix],
                                trainable=False)
    #################################################################################################################################

  0%|          | 0/2304 [00:00<?, ?it/s]




Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where


 36%|███▌      | 826/2304 [42:15<1:12:44,  2.95s/it]

In [ ]:
pd.DataFrame(result_list).sort_values(0, ascending=False)[0:10]

### Train final model

In [16]:
############### Clean up and rebuid objects; otherwise, tensor graph becomes larger and larger and requires more time. ##########
K.clear_session() # Will cause error: Tensor must from the same graph. Cause: Embedding layers changed. Solution: rebuild embedding layer.
##### Initialize session for reproducibility #####
np.random.seed(42)
rn.seed(12345)
session_conf = tf.ConfigProto(intra_op_parallelism_threads=1,
                              inter_op_parallelism_threads=1)
tf.set_random_seed(1234)
sess = tf.Session(graph=tf.get_default_graph(), config=session_conf)
K.set_session(sess)
##################################################
# Rebuid embedding layer.
embedding_layer = Embedding(input_dim=len(tokenizer.word_index)+1, # Size of vocabulary.
                            input_length=input_length, # Length of input, i.e., length of padded sequence.
                            output_dim=EMBEDDING_DIM, # Size of the vector space in which words will be embedded.
                            weights=[embedding_matrix],
                            trainable=False)
#################################################################################################################################

model = Sequential()
model.add(embedding_layer)
# model.add(Flatten())
model.add(Conv1D(32, 5, activation='softplus'))
model.add(GlobalMaxPool1D())
model.add(Dense(units=32, activation='softplus'))
model.add(Dense(units=16, activation='softplus'))
model.add(Dense(units=len(y_train[0]), activation='softplus'))
# compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['acc'])
# F1, precision, and recall removed. https://github.com/keras-team/keras/issues/5794
# fit the model
history=model.fit(x_train, y_train, validation_split=0.3, epochs=22, verbose=1)

Train on 138 samples, validate on 60 samples
Epoch 1/22
138/138 [==============================] - 0s 3ms/step - loss: 0.9108 - acc: 0.4130 - val_loss: 0.7422 - val_acc: 0.4167
Epoch 2/22
138/138 [==============================] - 0s 178us/step - loss: 0.7105 - acc: 0.4275 - val_loss: 0.6686 - val_acc: 0.6667
Epoch 3/22
138/138 [==============================] - 0s 155us/step - loss: 0.6690 - acc: 0.5797 - val_loss: 0.6626 - val_acc: 0.5833
Epoch 4/22
138/138 [==============================] - 0s 157us/step - loss: 0.6665 - acc: 0.5870 - val_loss: 0.6662 - val_acc: 0.5833
Epoch 5/22
138/138 [==============================] - 0s 157us/step - loss: 0.6574 - acc: 0.5870 - val_loss: 0.6634 - val_acc: 0.5833
Epoch 6/22
138/138 [==============================] - 0s 212us/step - loss: 0.6440 - acc: 0.5870 - val_loss: 0.6582 - val_acc: 0.5833
Epoch 7/22
138/138 [==============================] - 0s 166us/step - loss: 0.6275 - acc: 0.5870 - val_loss: 0.6528 - val_acc: 0.5833
Epoch 8/22
138/138 

In [ ]:
model.predict(x_pred)

## Traps.

- Tried GridSearchCV, suffers from the "tensor graph becomes larger and larger" problem.

<s>
    
```Python
def create_model(num_filters, kernel_size, conv_act, act_first, act_second, out_act):
    model = Sequential()
    model.add(embedding_layer)
    model.add(Conv1D(filters=num_filters, kernel_size=kernel_size, activation=conv_act))
    model.add(GlobalMaxPool1D())
    model.add(Dense(units=32, activation=act_first))
    model.add(Dense(units=16, activation=act_second))
    model.add(Dense(units=len(y_train[0]), activation=out_act))
    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['acc'])
    return model

from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import GridSearchCV
neural_network = KerasClassifier(build_fn=create_model, validation_split=0.3, verbose=0)
grid = GridSearchCV(estimator=neural_network, param_grid=hyperparameters)
grid_result = grid.fit(x_train, y_train, verbose=1)
```
    
</s>

- Give up trying multiprocessing: fail to send embedding layer (a pickle object) to threads.
    - possible solution: buid embedding layer on remote engines/threads.
        - Tried, looks like not working. Since it now only takes about 1.5 hours to complete the job, give up, move on.

<s>

```Python
import ipyparallel as ipp
c = ipp.Client()
print(c.ids)
dview = c[:]

dview['Sequential']=Sequential
dview['embedding_layer']=embedding_layer
dview['Conv1D']=Conv1D
dview['GlobalMaxPool1D']=GlobalMaxPool1D
dview['Dense']=Dense

@dview.parallel(block=True)
def grid_search(params):
    model = Sequential()
    model.add(embedding_layer)
    # model.add(Flatten())
    model.add(Conv1D(params[0], params[1], activation=params[2]))
    model.add(GlobalMaxPool1D())
    model.add(Dense(units=32, activation=params[3]))
    model.add(Dense(units=16, activation=params[4]))
    model.add(Dense(units=len(y_train[0]), activation=params[5]))
    # compile the model
    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['acc'])
    # F1, precision, and recall removed. https://github.com/keras-team/keras/issues/5794
    # fit the model
    history=model.fit(x_train, y_train, validation_split=0.3, epochs=100, verbose=0)
    return [[max(history.history['val_acc']), [num_filters, kernel_size, conv_act, act_32, act_16, out_act]]]

grid_search.map(param_list)

def grid_search(params):
    with tf.device('/gpu:0'): # Specify which GPU to use.
        model = Sequential()
        model.add(embedding_layer)
        # model.add(Flatten())
        model.add(Conv1D(params[0], params[1], activation=params[2]))
        model.add(GlobalMaxPool1D())
        model.add(Dense(units=32, activation=params[3]))
        model.add(Dense(units=16, activation=params[4]))
        model.add(Dense(units=len(y_train[0]), activation=params[5]))
        # compile the model
        model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['acc'])
        # F1, precision, and recall removed. https://github.com/keras-team/keras/issues/5794
        # fit the model
        history=model.fit(x_train, y_train, validation_split=0.3, epochs=100, verbose=0)
    return [[max(history.history['val_acc']), [num_filters, kernel_size, conv_act, act_32, act_16, out_act]]]

from multiprocessing import Pool
# if __name__== "__main__":
p=Pool(10)
result_list=p.map(grid_search, param_list)
```

</s>